### PR TITLE
[diem-transactional-tests] add support for block metadata transactions + migrate some tests -- 1/N

### PR DESCRIPTION
**Pull request by vgao1996**
This PR contains the following stack of commits:
1. Support for block metadata transactions.
2. Allowing transaction expiration time to be specified as an extra arg.
3. Migrating old block functional tests. (All manually inspected/rewritten, none deleted)

### CODE REVIEW

File: diem-move/diem-transactional-test-harness/src/diem_test_harness.rs [hunk 1]
```diff
@@ -448,9 +503,81 @@ impl<'a> DiemTestAdapter<'a> {
...
+                parameters.gas_unit_price,
+                parameters.gas_currency_code,
+                parameters.expiration_timestamp_secs,
+                ChainId::test(),
+            )
+            .sign(&GENESIS_KEYPAIR.0, GENESIS_KEYPAIR.1.clone())
+            .unwrap()
+            .into_inner();
+        self.run_transaction(Transaction::UserTransaction(txn))
+            .expect("Failed to add validator to validator set. This should not happen.");
```
Comment (vgao1996) [inline diem-move/diem-transactional-test-harness/src/diem_test_harness.rs:580]: Yeah, it turned out the validator account creation procedure I previously implemented was barely sufficient. It's missing the subsequent steps (creating and setting the operator, registering the validator config and adding the validator to the validator set) so I had to fix it here.

**Comment by modocache** [comment]
I'd like to review this, but to do so properly I think I'd need to learn more about what block metadata transactions are, and the motivation for providing expiration time in an argument. Is there a link to a description of the work being done here, or could you explain it to me here (or in the commit messages)?

**Comment by vgao1996** [comment]
@modocache Thanks for helping! I'm not ware of existing docs explaining this concept well so I'll say it in my own words. 
- In the real world, Diem transactions are grouped into "blocks" before getting executed.
- A special transaction, called the "block metadata transaction" or ("block prologue") is inserted at the beginning of every block.
- This transaction updates some of the un-chain metadata, including a global clock. This is how transactions can gain access of the concept of "time" during execution.
- The block metadata transaction must be proposed by a validator (the one leading consensus?).
- Transactional tests are a simplified/idealized simulation of the real workflow, so we need to support block metadata transactions.

File: diem-move/diem-framework/core/tests/DiemBlockTests.move [hunk 2]
```diff
@@ -0,0 +1,13 @@
+#[test_only]
+module DiemFramework::DiemBlockTests {
+    use DiemFramework::Genesis;
+    use DiemFramework::DiemBlock;
+
+    // REVIEW: is this abort code correct?
```
Comment (tnowacki) [inline diem-move/diem-framework/core/tests/DiemBlockTests.move:6]: @tzakian ?

File: diem-move/diem-framework/core/transactional-tests/block/block_prologue.move [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+//#     --signers DiemRoot
+script{
+use DiemFramework::DiemTimestamp;
+
+fun main() {
+    assert!(DiemTimestamp::now_microseconds() != 2000000, 77);
+}
+}
+
+// REVIEW: what is this doing? Shall we delete it?
```
Comment (tnowacki) [inline diem-move/diem-framework/core/transactional-tests/block/block_prologue.move:27]: What is the 20 there? Is it resetting the time to 20? Does this interact with the block prologue in some way?

File: diem-move/diem-transactional-test-harness/src/diem_test_harness.rs [hunk 4]
```diff
@@ -94,6 +100,9 @@ struct DiemRunArgs {
 
     #[structopt(long = "--admin-script")]
     admin_script: bool,
+
+    #[structopt(short = "t", long = "expiration")]
+    expiration_time: Option<u64>,
```
Comment (tnowacki) [inline diem-move/diem-transactional-test-harness/src/diem_test_harness.rs:105]: Ditto

File: diem-move/diem-transactional-test-harness/src/diem_test_harness.rs [hunk 5]
```diff
@@ -412,31 +435,63 @@ impl<'a> DiemTestAdapter<'a> {
...
-        Ok(())
     }
 
     /// Create a validator account with the given credentials.
     ///
     /// Note: this does not add it to the named address or private key mappings.
     /// That needs to be done separately.
+    ///
+    /// REVIEW: currently it is impossible to use the same validator account as the operator.
+    /// Check if this is intended.
```
Comment (tnowacki) [inline diem-move/diem-transactional-test-harness/src/diem_test_harness.rs:446]: I don't understand the comment, maybe @tzakian or @emmazzz can help

File: diem-move/diem-transactional-test-harness/src/diem_test_harness.rs [hunk 6]
```diff
@@ -830,11 +994,18 @@ impl<'a> MoveTestAdapter<'a> for DiemTestAdapter<'a> {
...
-    ) -> Result<Option<String>> {
-        unreachable!()
+    fn handle_subcommand(&mut self, input: TaskInput<Self::Subcommand>) -> Result<Option<String>> {
+        match input.command {
+            DiemSubCommand::BlockCommand(block_cmd) => {
+                let proposer = self.compiled_state().resolve_address(&block_cmd.proposer);
+                let metadata =
+                    BlockMetadata::new(HashValue::zero(), 0, block_cmd.time, vec![], proposer);
+
+                let output = self.run_transaction(Transaction::BlockMetadata(metadata))?;
```
Comment (tnowacki) [inline diem-move/diem-transactional-test-harness/src/diem_test_harness.rs:1004]: Didn't realize this was truly its own transaction type. Seems like a good use case for a subcommand then

File: diem-move/diem-transactional-test-harness/tests/diem_test_harness/block_metadata.exp [hunk 7]
```diff
@@ -0,0 +1,5 @@
+processed 2 tasks
+
+task 1 'block'. lines 3-3:
+Events:
+ContractEvent { key: EventKey(11000000000000000000000000000000000000000a550c18), index: 0, type: Struct(StructTag { address: 00000000000000000000000000000001, module: Identifier("DiemBlock"), name: Identifier("NewBlockEvent"), type_params: [] }), event_data: "0000000000000000c5e34b925cf6875fec02d4b77adbd2d6006400000000000000" }
```
Comment (tnowacki) [inline diem-move/diem-transactional-test-harness/tests/diem_test_harness/block_metadata.exp:5]: The event output is rather difficult to read. Could you write a custom printer if it isn't too much work?

Comment (vgao1996) [reply diem-move/diem-transactional-test-harness/tests/diem_test_harness/block_metadata.exp:5]: Will do, but probably in a separate commit. (I have a bunch coming in the way so there will be chances for me to insert this soon.)

File: diem-move/diem-framework/core/transactional-tests/block/block_prologue.move (hunk 3 above)
Comment (vgao1996) [reply diem-move/diem-framework/core/transactional-tests/block/block_prologue.move:27]: The block prologue sets the timestamp, and this seem to do the same thing. Although I really don't think we should do this via an admin script, hence I'm asking if we should delete or rewrite it.

File: diem-move/diem-transactional-test-harness/src/diem_test_harness.rs [hunk 8]
```diff
@@ -84,6 +87,9 @@ struct TransactionParameters {
 struct DiemPublishArgs {
     #[structopt(short = "k", long = "private-key", parse(try_from_str = RawPrivateKey::parse))]
     private_key: Option<RawPrivateKey>,
+
+    #[structopt(short = "t", long = "expiration")]
+    expiration_time: Option<u64>,
```
[the remark this answers is not in the document]
Comment (vgao1996) [reply diem-move/diem-transactional-test-harness/src/diem_test_harness.rs:92]: Every transaction has an expiration time. If this ends up being earlier than the timestamp set in the block prologue, then the transaction gets rejected. See my explanation about block metadata transactions to Brian for more details.

File: diem-move/diem-transactional-test-harness/src/diem_test_harness.rs (hunk 4 above)
Comment (vgao1996) [reply diem-move/diem-transactional-test-harness/src/diem_test_harness.rs:105]: Same as above.

File: diem-move/diem-transactional-test-harness/src/diem_test_harness.rs (hunk 5 above)
Comment (vgao1996) [reply diem-move/diem-transactional-test-harness/src/diem_test_harness.rs:446]: Here's the gist: I found it impossible to set the operator of a validator to itself via `public(script) fun set_validator_operator`, which seem to contradict with the comment in the following code:
```
    /// Sets up the initial validator set for the Diem network.
    /// The validator "owner" accounts, their UTF-8 names, and their authentication
    /// keys are encoded in the `owners`, `owner_names`, and `owner_auth_key` vectors.
    /// Each validator signs consensus messages with the private key corresponding to the Ed25519
    /// public key in `consensus_pubkeys`.
    /// Each validator owner has its operation delegated to an "operator" (which may be
    /// the owner). The operators, their names, and their authentication keys are encoded
    /// in the `operators`, `operator_names`, and `operator_auth_keys` vectors.
    /// Finally, each validator must specify the network address
    /// (see diem/types/src/network_address/mod.rs) for itself and its full nodes.
    fun create_initialize_owners_operators(
        dr_account: signer,
        owners: vector<signer>,
        owner_names: vector<vector<u8>>,
        owner_auth_keys: vector<vector<u8>>,
        consensus_pubkeys: vector<vector<u8>>,
        operators: vector<signer>,
        operator_names: vector<vector<u8>>,
        operator_auth_keys: vector<vector<u8>>,
        validator_network_addresses: vector<vector<u8>>,
        full_node_network_addresses: vector<vector<u8>>,
    ) {
        let num_owners = Vector::length(&owners);
        let num_owner_names = Vector::length(&owner_names);
        assert!(num_owners == num_owner_names, 0);
        let num_owner_keys = Vector::length(&owner_auth_keys);
        assert!(num_owner_names == num_owner_keys, 0);
        let num_operators = Vector::length(&operators);
        assert!(num_owner_keys == num_operators, 0);
        let num_operator_names = Vector::length(&operator_names);
        assert!(num_operators == num_operator_names, 0);
        let num_operator_keys = Vector::length(&operator_auth_keys);
        assert!(num_operator_names == num_operator_keys, 0);
        let num_validator_network_addresses = Vector::length(&validator_network_addresses);
        assert!(num_operator_keys == num_validator_network_addresses, 0);
        let num_full_node_network_addresses = Vector::length(&full_node_network_addresses);
        assert!(num_validator_network_addresses == num_full_node_network_addresses, 0);

        let i = 0;
        let dummy_auth_key_prefix = x"00000000000000000000000000000000";
        while (i < num_owners) {
            let owner = Vector::borrow(&owners, i);
            let owner_address = Signer::address_of(owner);
            let owner_name = *Vector::borrow(&owner_names, i);
            // create each validator account and rotate its auth key to the correct value
            DiemAccount::create_validator_account(
                &dr_account, owner_address, copy dummy_auth_key_prefix, owner_name
            );

            let owner_auth_key = *Vector::borrow(&owner_auth_keys, i);
            let rotation_cap = DiemAccount::extract_key_rotation_capability(owner);
            DiemAccount::rotate_authentication_key(&rotation_cap, owner_auth_key);
            DiemAccount::restore_key_rotation_capability(rotation_cap);

            let operator = Vector::borrow(&operators, i);
            let operator_address = Signer::address_of(operator);
            let operator_name = *Vector::borrow(&operator_names, i);
            // create the operator account + rotate its auth key if it does not already exist
            if (!DiemAccount::exists_at(operator_address)) {
                DiemAccount::create_validator_operator_account(
                    &dr_account, operator_address, copy dummy_auth_key_prefix, copy operator_name
                );
                let operator_auth_key = *Vector::borrow(&operator_auth_keys, i);
                let rotation_cap = DiemAccount::extract_key_rotation_capability(operator);
                DiemAccount::rotate_authentication_key(&rotation_cap, operator_auth_key);
                DiemAccount::restore_key_rotation_capability(rotation_cap);
            };
            // assign the operator to its validator
            assert!(ValidatorOperatorConfig::get_human_name(operator_address) == operator_name, 0);
            ValidatorConfig::set_operator(owner, operator_address);

            // use the operator account set up the validator config
            let validator_network_address = *Vector::borrow(&validator_network_addresses, i);
            let full_node_network_address = *Vector::borrow(&full_node_network_addresses, i);
            let consensus_pubkey = *Vector::borrow(&consensus_pubkeys, i);
            ValidatorConfig::set_config(
                operator,
                owner_address,
                consensus_pubkey,
                validator_network_address,
                full_node_network_address
            );

            // finally, add this validator to the validator set
            DiemSystem::add_validator(&dr_account, owner_address);

            i = i + 1;
        }
    }
```
Here `ValidatorOperatorConfig::get_human_name(operator_address) ` will abort if `if (!DiemAccount::exists_at(operator_address)) {` is not taken. 

Nevertheless, this doesn't prevent us from implementing the proper validator creation process here...

**Review by tnowacki — DISMISSED** [review]

**Pull request merged**